CRM/Contribute/Form/CancelSubscription.php

```diff
@@ -220,8 +220,13 @@ public function postProcess() {
 
     if (CRM_Utils_Array::value('send_cancel_request', $params) == 1) {
       $cancelParams = ['subscriptionId' => $this->_subscriptionDetails->subscription_id];
-      $this->_paymentProcessorObj->setContributionRecurID($this->contributionRecurID);
-      $cancelSubscription = $this->_paymentProcessorObj->cancelSubscription($message, $cancelParams);
```

Comment:
I think we should add a new function to CRM_Core_Payment 'doCancelSubscription' & call it here. That function would call cancelSubscription but ensure that an exception is thrown if need be .

I'm OK with your changes but this also helps us resolve another issue of this function

```
   /**
   * Does this processor support cancelling recurring contributions through code.
   *
   * If the processor returns true it must be possible to take action from within CiviCRM
   * that will result in no further payments being processed. In the case of token processors (e.g
   * IATS, eWay) updating the contribution_recur table is probably sufficient.
   *
   * @return bool
   */
  protected function supportsCancelRecurring() {
    return method_exists(CRM_Utils_System::getClassName($this), 'cancelSubscription');
  }
```

Which has caused me pain before (although I can't fully remember the details now)